### PR TITLE
e2e: retry deploying on API-server timeouts

### PR DESCRIPTION
**Pull request by nixpanic**
The upgrade-tests-cephfs fails relative regularly with the following
error during intial deployment:

    timeout waiting for deployment csi-cephfsplugin-provisioner with error error waiting for deployment "csi-cephfsplugin-provisioner" status to match expectation: etcdserver: request timed out

By detecting if the API-server returned ErrWaitTimeout, the test does
not need to abort, but can wait for completion. PollImmediate() will
still return ErrWaitTimeout once the timeout elapsed.

### CODE REVIEW

**Comment by nixpanic** [comment]
Added `[DNM]` so that multiple `/retest ci/centos/upgrade-tests-cephfs` can be done to verify that the test is more stable.

File: e2e/pod.go [hunk 1]
```diff
@@ -59,6 +59,11 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
 	err = wait.PollImmediate(poll, timeout, func() (bool, error) {
 		deployment, err = c.AppsV1().Deployments(ns).Get(context.TODO(), name, metav1.GetOptions{})
 		if err != nil {
+			if errors.Is(err, wait.ErrWaitTimeout) {
```
Comment (Madhu-1) [inline e2e/pod.go:62]: can you try with
```go
if testutils.IsRetryableAPIError(err) {
				return false, nil
}
```

This check is already present in `waitForDaemonSets` can we include this in other places also where ever we are pooling for the response

Comment (nixpanic) [reply e2e/pod.go:62]: sure!

**Comment by nixpanic** [comment]
> /retest ci/centos/upgrade-tests-cephfs

The job passed 2x before, and this failure is due to some infrastructure error?!

    + dnf -y install git podman make
    ./prepare.sh: line 81: dnf: command not found

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
ci/centos/upgrade-tests-rbd failed in a timeout before running the actual test

**Comment by nixpanic** [comment]
**ci/centos/upgrade-tests-cephfs** failed with:

    Oct 23 13:16:34.222: timeout waiting for deployment csi-cephfsplugin-provisioner with error error waiting for deployment "csi-cephfsplugin-provisioner" status to match expectation: etcdserver: request timed out

This suggests the current PR does not address the issue.

**Comment by nixpanic** [comment]
@mergifyio rebase

**Comment by nixpanic** [comment]
Hmm **ci/centos/mini-e2e/k8s-1.19.2** failed due to some timeout that isn't identified yet. Not a problem that this PR attempts to address.

**Comment by nixpanic** [comment]
The tests have failed once with the additional retry function. Maybe it catches most of the issues, just not all? I guess this is good to get merged. @Madhu-1 what do you think?

**Comment by Yuggupta27** [comment]
> The tests have failed once with the additional retry function. Maybe it catches most of the issues, just not all? I guess this is good to get merged. @Madhu-1 what do you think?

Was the error in the failed run of etcd timeout? Or any other error?

**Comment by Madhu-1** [comment]
> **ci/centos/upgrade-tests-cephfs** failed with:
> 
> ```
> Oct 23 13:16:34.222: timeout waiting for deployment csi-cephfsplugin-provisioner with error error waiting for deployment "csi-cephfsplugin-provisioner" status to match expectation: etcdserver: request timed out
> ```
> 
> This suggests the current PR does not address the issue.

@nixpanic is this still true?

**Comment by nixpanic** [comment]
> > The tests have failed once with the additional retry function. Maybe it catches most of the issues, just not all? I guess this is good to get merged. @Madhu-1 what do you think?
> 
> Was the error in the failed run of etcd timeout? Or any other error?

The etcdserver timeout, see https://github.com/ceph/ceph-csi/pull/1626#issuecomment-715314929

**Comment by nixpanic** [comment]
> > **ci/centos/upgrade-tests-cephfs** failed with:
> > ```
> > Oct 23 13:16:34.222: timeout waiting for deployment csi-cephfsplugin-provisioner with error error waiting for deployment "csi-cephfsplugin-provisioner" status to match expectation: etcdserver: request timed out
> > ```
> > 
> > 
> > This suggests the current PR does not address the issue.
> 
> @nixpanic is this still true?

Yes, that is the only time it failed with the error. I only added additional logging afterwards to see why the error is not caught. But since then, it did not reproduce anymore.

File: e2e/pod.go [hunk 2]
```diff
@@ -59,6 +59,13 @@ func waitForDeploymentComplete(name, ns string, c kubernetes.Interface, t int) e
 	err = wait.PollImmediate(poll, timeout, func() (bool, error) {
 		deployment, err = c.AppsV1().Deployments(ns).Get(context.TODO(), name, metav1.GetOptions{})
 		if err != nil {
+			if testutils.IsRetryableAPIError(err) {
+				// hide API-server timeouts, so that PollImmediate() retries
+				e2elog.Logf("deployment error: %v", err)
+				return false, nil
+			}
+			// FIXME: this is for debugging etcdserver timeouts only
+			e2elog.Logf("[DEBUG] deployment check failed: %v (%T)", err, err)
```
Comment (Madhu-1) [inline e2e/pod.go:68]: remove Fixme?

Comment (nixpanic) [reply e2e/pod.go:68]: Well, it failed once, so we might still need to fix things here.

Comment (nixpanic) [reply e2e/pod.go:68]: And it actually failed the last run :partying_face: 

    Oct 27 11:27:25.860: INFO: [DEBUG] deployment check failed: etcdserver: request timed out (*errors.StatusError)

Comment (nixpanic) [reply e2e/pod.go]: This kind of error is now caught as well, and a retry will be done.

**Comment by nixpanic** [comment]
@mergifyio rebase

**Review by Madhu-1 — APPROVED** [review]

**Comment by Madhu-1** [comment]
>Resetting dropped connection: admin.ci.centos.org
Resetting dropped connection: admin.ci.centos.org
+---------+-----------+-------------+---------+------------+---------------+----------+--------+------+----------------+--------------+-----------+--------------+--------+
| host_id | hostname  | ip_address  | chassis | used_count | current_state | comment  | distro | rel  | centos_version | architecture | node_pool | console_port | flavor |
+---------+-----------+-------------+---------+------------+---------------+----------+--------+------+----------------+--------------+-----------+--------------+--------+
|      90 | n26.pufty | 172.19.3.90 | pufty   |       5042 | Deployed      | f71bc4c5 | None   | None | 8              | x86_64       |         1 |         2250 | None   |
+---------+-----------+-------------+---------+------------+---------------+----------+--------+------+----------------+--------------+-----------+--------------+--------+
[Pipeline] }
[Pipeline] // stage
[Pipeline] }
[Pipeline] // node
[Pipeline] End of Pipeline
[Checks API] No suitable checks publisher found.

CI failed when deploying Rook

**Pull request merged**